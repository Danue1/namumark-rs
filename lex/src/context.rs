pub struct Context<'lex> {
    source: &'lex str,
    index: usize,
}

impl<'lex> Context<'lex> {
    pub const fn new(source: &'lex str) -> Self {
        Self { source, index: 0 }
    }

    pub const fn index(&self) -> usize {
        self.index
    }

    pub fn peek(&self) -> Option<char> {
        self.source.get(self.index..).and_then(|s| s.chars().next())
    }

    pub fn bump(&mut self) -> Option<char> {
        let c = self.peek()?;
        self.index += c.len_utf8();
        Some(c)
    }

    #[inline]
    pub fn slice(&self, start: usize) -> String {
        self.source[start..self.index].to_owned()
    }
}
