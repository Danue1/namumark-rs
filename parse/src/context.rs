use rowan::{Checkpoint, GreenNodeBuilder};
use syntax_kind::{SyntaxKind, SyntaxNode};

#[derive(Debug)]
pub struct Context {
    tokens: Vec<(SyntaxKind, String)>,
    builder: GreenNodeBuilder<'static>,
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum State {
    Continue,
    Stop,
}

#[allow(unused)]
impl Context {
    pub fn new(tokens: impl Iterator<Item = (SyntaxKind, String)>) -> Self {
        let mut tokens: Vec<_> = tokens.collect();
        tokens.reverse();
        Self {
            tokens,
            builder: GreenNodeBuilder::new(),
        }
    }

    pub fn peek(&self) -> SyntaxKind {
        match self.tokens.last() {
            Some((kind, _)) => *kind,
            None => SyntaxKind::EOF,
        }
    }

    pub fn peek_str(&self) -> &str {
        match self.tokens.last() {
            Some((_, text)) => text.as_str(),
            None => "",
        }
    }

    pub fn nth(&self, size: usize) -> SyntaxKind {
        match self.tokens.get(self.tokens.len() - size) {
            Some((kind, _)) => *kind,
            None => SyntaxKind::EOF,
        }
    }

    pub fn bump(&mut self) {
        if let Some((kind, text)) = self.tokens.pop() {
            self.builder.token(kind.into(), text.as_str());
        }
    }

    pub fn skip_whitespace(&mut self) {
        while self.peek() == SyntaxKind::WHITESPACE {
            self.bump();
        }
    }

    #[inline]
    pub fn is_empty(&self) -> bool {
        self.tokens.is_empty()
    }

    #[inline]
    pub fn checkpoint(&self) -> Checkpoint {
        self.builder.checkpoint()
    }

    #[inline]
    pub fn start_node_at(&mut self, checkpoint: Checkpoint, kind: SyntaxKind) {
        self.builder.start_node_at(checkpoint, kind.into());
    }

    #[inline]
    pub fn start_node(&mut self, kind: SyntaxKind) {
        self.builder.start_node(kind.into());
    }

    #[inline]
    pub fn finish_node(&mut self) {
        self.builder.finish_node();
    }

    #[inline]
    pub fn finish(self) -> SyntaxNode {
        SyntaxNode::new_root(self.builder.finish())
    }
}

#[macro_export]
macro_rules! one_of {
    ($head:expr,) => {
        $head
    };
    ($head:expr, $($tail:expr,)+) => {
        if $head == crate::State::Continue {
            one_of!($($tail,)+)
        } else {
            crate::State::Stop
        }
    };
}
