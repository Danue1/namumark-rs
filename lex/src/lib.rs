mod context;

use context::Context;
use syntax_kind::SyntaxKind;

pub fn lex(source: &str) -> impl Iterator<Item = (SyntaxKind, String)> + '_ {
    let mut context = Context::new(source);
    std::iter::from_fn(move || {
        let start = context.index();
        let token = lex_token(&mut context)?;
        let text = context.slice(start);
        Some((token, text))
    })
}

fn lex_token(context: &mut Context) -> Option<SyntaxKind> {
    context.bump().map(|c| {
        if let Some(kind) = punctuation(c) {
            kind
        } else if c.is_numeric() {
            numeric(context)
        } else {
            text(context)
        }
    })
}

const fn punctuation(c: char) -> Option<SyntaxKind> {
    let kind = match c {
        '(' => SyntaxKind::LEFT_PAREN,
        ')' => SyntaxKind::RIGHT_PAREN,
        '{' => SyntaxKind::LEFT_BRACE,
        '}' => SyntaxKind::RIGHT_BRACE,
        '[' => SyntaxKind::LEFT_BRACKET,
        ']' => SyntaxKind::RIGHT_BRACKET,
        '<' => SyntaxKind::LEFT_CHEVRON,
        '>' => SyntaxKind::RIGHT_CHEVRON,
        ',' => SyntaxKind::COMMA,
        '.' => SyntaxKind::DOT,
        '-' => SyntaxKind::MINUS,
        '+' => SyntaxKind::PLUS,
        ':' => SyntaxKind::COLON,
        ';' => SyntaxKind::SEMICOLON,
        '=' => SyntaxKind::EQUAL,
        '/' => SyntaxKind::SLASH,
        '\\' => SyntaxKind::BACKSLASH,
        '*' => SyntaxKind::STAR,
        '#' => SyntaxKind::HASH,
        '@' => SyntaxKind::AT,
        '$' => SyntaxKind::DOLLAR,
        '%' => SyntaxKind::PERCENT,
        '^' => SyntaxKind::CARET,
        '&' => SyntaxKind::AMPERSAND,
        '|' => SyntaxKind::PIPE,
        '~' => SyntaxKind::TILDE,
        '_' => SyntaxKind::UNDERSCORE,
        '?' => SyntaxKind::QUESTION,
        '!' => SyntaxKind::EXCLAMATION,
        '\'' => SyntaxKind::SINGLE_QUOTE,
        '"' => SyntaxKind::DOUBLE_QUOTE,
        '`' => SyntaxKind::BACKTICK,
        ' ' => SyntaxKind::WHITESPACE,
        '\t' => SyntaxKind::TAB,
        '\n' => SyntaxKind::NEW_LINE,
        _ => return None,
    };
    Some(kind)
}

fn numeric(context: &mut Context) -> SyntaxKind {
    while let Some(c) = context.peek() {
        if !c.is_numeric() {
            break;
        }
        context.bump();
    }
    SyntaxKind::NUMERIC
}

fn text(context: &mut Context) -> SyntaxKind {
    while let Some(c) = context.peek() {
        if punctuation(c).is_some() || c.is_numeric() {
            break;
        }
        context.bump();
    }
    SyntaxKind::TEXT
}

#[cfg(test)]
mod tests {
    use syntax_kind::SyntaxKind;

    macro_rules! assert_snapshot {
        ($source:expr) => {
            let source = $source;
            let tokens: Vec<(SyntaxKind, String)> = super::lex(source).collect();
            insta::assert_debug_snapshot!(tokens);
        };
    }

    #[test]
    fn punctuation() {
        assert_snapshot!("(){}[]<>,.-+:;=/\\*#@$%^&|~_?!'\"` \t\n");
    }

    #[test]
    fn numeric() {
        assert_snapshot!("1234567890");
    }

    #[test]
    fn text_alphabet() {
        assert_snapshot!("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ");
    }

    #[test]
    fn text_hangul() {
        assert_snapshot!("가나다라마바사아자차카타파하");
    }
}
