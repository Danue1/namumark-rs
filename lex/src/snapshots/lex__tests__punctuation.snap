---
source: lex/src/lib.rs
expression: tokens
---
[
    (
        LEFT_PAREN,
        "(",
    ),
    (
        RIGHT_PAREN,
        ")",
    ),
    (
        LEFT_BRACE,
        "{",
    ),
    (
        RIGHT_BRACE,
        "}",
    ),
    (
        LEFT_BRACKET,
        "[",
    ),
    (
        RIGHT_BRACKET,
        "]",
    ),
    (
        LEFT_CHEVRON,
        "<",
    ),
    (
        RIGHT_CHEVRON,
        ">",
    ),
    (
        COMMA,
        ",",
    ),
    (
        DOT,
        ".",
    ),
    (
        MINUS,
        "-",
    ),
    (
        PLUS,
        "+",
    ),
    (
        COLON,
        ":",
    ),
    (
        SEMICOLON,
        ";",
    ),
    (
        EQUAL,
        "=",
    ),
    (
        SLASH,
        "/",
    ),
    (
        BACKSLASH,
        "\\",
    ),
    (
        STAR,
        "*",
    ),
    (
        HASH,
        "#",
    ),
    (
        AT,
        "@",
    ),
    (
        DOLLAR,
        "$",
    ),
    (
        PERCENT,
        "%",
    ),
    (
        CARET,
        "^",
    ),
    (
        AMPERSAND,
        "&",
    ),
    (
        PIPE,
        "|",
    ),
    (
        TILDE,
        "~",
    ),
    (
        UNDERSCORE,
        "_",
    ),
    (
        QUESTION,
        "?",
    ),
    (
        EXCLAMATION,
        "!",
    ),
    (
        SINGLE_QUOTE,
        "'",
    ),
    (
        DOUBLE_QUOTE,
        "\"",
    ),
    (
        BACKTICK,
        "`",
    ),
    (
        WHITESPACE,
        " ",
    ),
    (
        TAB,
        "\t",
    ),
    (
        NEW_LINE,
        "\n",
    ),
]
